File: packages/java/core/src/test/java/dev/configdiscovery/ConfigDiscoveryTest.java
package dev.configdiscovery;

import static dev.configdiscovery.TestTrees.tree;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import java.nio.file.Path;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

class ConfigDiscoveryTest {

    private static final ObjectMapper MAPPER = new ObjectMapper();

    private static LoadOptions.Builder options(Path root) {
        return LoadOptions.builder()
                .cwd(root.resolve("project"))
                .home(root.resolve("home"))
                .env(Map.of())
                .warningHandler(message -> {});
    }

    @Test
    void localOnly(@TempDir Path temp) {
        Path root = tree(
                temp,
                "project/.git", "",
                "project/.config/mytool/config.toml", "[log]\nlevel = \"debug\"\n",
                "home/", "");
        Loaded loaded = ConfigDiscovery.load("mytool", options(root).build());
        assertEquals("debug", loaded.config().get("log").get("level").asText());
        assertTrue(loaded.found());
        assertEquals(1, loaded.sources().size());
        assertEquals(2, loaded.sources().get(0).precedence());
    }

    @Test
    void userOnly(@TempDir Path temp) {
        Path root = tree(temp, "project/.git", "", "home/.config/mytool/config.toml", "[log]\nlevel = \"info\"\n");
        Loaded loaded = ConfigDiscovery.load("mytool", options(root).build());
        assertEquals(1, loaded.sources().get(0).precedence());
    }

    @Test
    void neitherPresentIsNotAnError(@TempDir Path temp) {
        Path root = tree(temp, "project/.git", "", "home/", "");
        Loaded loaded = ConfigDiscovery.load("mytool", options(root).build());
        assertFalse(loaded.found());
        assertTrue(loaded.config().isEmpty());
        assertTrue(loaded.sources().isEmpty());
    }

    @Test
    void defaultsSurviveButDoNotSetFound(@TempDir Path temp) {
        Path root = tree(temp, "project/.git", "", "home/", "");
        ObjectNode defaults = MAPPER.createObjectNode();
        defaults.putObject("log").put("level", "warn");
        Loaded loaded = ConfigDiscovery.load("mytool", options(root).defaults(defaults).build());
        assertEquals("warn", loaded.config().get("log").get("level").asText());
        assertFalse(loaded.found());
        assertEquals("defaults", loaded.sources().get(0).format());
    }

    @Test
    void threeLayersConflictAndTheLosersAreStillReported(@TempDir Path temp) {
        Path root = tree(
                temp,
                "project/.git", "",
                "project/.config/mytool/config.toml", "[log]\nlevel = \"debug\"\n",
                "home/.config/mytool/config.toml", "[log]\nlevel = \"info\"\n");
        Loaded loaded = ConfigDiscovery.load(
                "mytool",
                options(root)
                        .env(Map.of("MYTOOL_LOG__LEVEL", "trace", "MYTOOL_PORT", "5432"))
                        .build());
        assertEquals("trace", loaded.config().get("log").get("level").asText());
        assertEquals(5432, loaded.config().get("port").asInt());
        assertEquals(List.of(1, 2, 4), loaded.sources().stream().map(Source::precedence).toList());
    }

    @Test
    void overridesWinOverEverything(@TempDir Path temp) {
        Path root = tree(
                temp,
                "project/.git", "",
                "project/.config/mytool/config.toml", "[log]\nlevel = \"debug\"\n",
                "home/", "");
        ObjectNode overrides = MAPPER.createObjectNode();
        overrides.putObject("log").put("level", "silent");
        Loaded loaded = ConfigDiscovery.load(
                "mytool",
                options(root)
                        .env(Map.of("MYTOOL_LOG__LEVEL", "trace"))
                        .overrides(overrides)
                        .build());
        assertEquals("silent", loaded.config().get("log").get("level").asText());
        assertEquals(5, loaded.sources().get(loaded.sources().size() - 1).precedence());
    }

    @Test
    void aPackageNameWithASeparatorIsRejected() {
        assertThrows(IllegalArgumentException.class, () -> ConfigDiscovery.load("../evil"));
    }

    @Test
    void everyErrorKindIsReachable(@TempDir Path temp) {
        Path malformed = tree(
                temp.resolve("a"), "project/.git", "", "project/.config/mytool/config.toml", "[log\n", "home/", "");
        assertEquals(
                ConfigException.Kind.MALFORMED,
                assertThrows(ConfigException.class, () -> ConfigDiscovery.load("mytool", options(malformed).build()))
                        .kind());

        Path duplicate = tree(
                temp.resolve("b"),
                "project/.git", "",
                "project/.config/mytool/config.yaml", "a: 1\n",
                "project/.config/mytool/config.yml", "a: 2\n",
                "home/", "");
        assertEquals(
                ConfigException.Kind.DUPLICATE_FORMAT,
                assertThrows(ConfigException.class, () -> ConfigDiscovery.load("mytool", options(duplicate).build()))
                        .kind());

        Path unreadable = tree(
                temp.resolve("c"), "project/.git", "", "project/.config/mytool/config.toml/", "", "home/", "");
        assertEquals(
                ConfigException.Kind.UNREADABLE,
                assertThrows(ConfigException.class, () -> ConfigDiscovery.load("mytool", options(unreadable).build()))
                        .kind());

        // NOT_FOUND exists so a caller can name the condition (SPEC section 5), not so it can be
        // thrown: an empty tree is a success.
        Path empty = tree(temp.resolve("d"), "project/.git", "", "home/", "");
        assertFalse(ConfigDiscovery.load("mytool", options(empty).build()).found());
        assertEquals("not-found", ConfigException.Kind.NOT_FOUND.wireName());
        // UNKNOWN_KEY and VALIDATION are reachable through binding; see BindingTest.
    }
}

File: packages/golang/configdiscovery_test.go
package configdiscovery

import (
	"encoding/json"
	"errors"
	"path/filepath"
	"reflect"
	"strings"
	"testing"
)

func asConfigError(err error, target **ConfigError) bool { return errors.As(err, target) }

func loadIn(t *testing.T, root string, opts ...Option) (*Loaded, error) {
	t.Helper()
	base := []Option{
		WithCwd(filepath.Join(root, "project")),
		WithHome(filepath.Join(root, "home")),
		WithEnv(map[string]string{}),
		WithWarningHandler(func(string) {}),
	}
	return Load("mytool", append(base, opts...)...)
}

func TestLoadLocalOnly(t *testing.T) {
	root := tree(t, map[string]string{
		"project/.git":                       "",
		"project/.config/mytool/config.toml": "[log]\nlevel = \"debug\"\n",
		"home/":                              "",
	})
	result, err := loadIn(t, root)
	if err != nil {
		t.Fatal(err)
	}
	if !reflect.DeepEqual(result.Config, map[string]any{"log": map[string]any{"level": "debug"}}) {
		t.Fatalf("got %#v", result.Config)
	}
	if !result.Found || len(result.Sources) != 1 || result.Sources[0].Precedence != 2 {
		t.Fatalf("got found=%v sources=%#v", result.Found, result.Sources)
	}
}

func TestLoadUserOnly(t *testing.T) {
	root := tree(t, map[string]string{
		"project/.git":                    "",
		"home/.config/mytool/config.toml": "[log]\nlevel = \"info\"\n",
	})
	result, err := loadIn(t, root)
	if err != nil {
		t.Fatal(err)
	}
	if result.Sources[0].Precedence != 1 {
		t.Fatalf("got %#v", result.Sources)
	}
}

func TestLoadNeitherPresentIsNotAnError(t *testing.T) {
	root := tree(t, map[string]string{"project/.git": "", "home/": ""})
	result, err := loadIn(t, root)
	if err != nil {
		t.Fatalf("nothing found must not be an error: %v", err)
	}
	if result.Found || len(result.Config) != 0 || len(result.Sources) != 0 {
		t.Fatalf("got %#v", result)
	}
	// SPEC section 7 and CANONICAL: an empty list is [], never null.
	encoded, err := json.Marshal(result)
	if err != nil {
		t.Fatal(err)
	}
	if !strings.Contains(string(encoded), `"sources":[]`) {
		t.Fatalf("empty sources must marshal as []: %s", encoded)
	}
}

func TestLoadDefaultsDoNotSetFound(t *testing.T) {
	root := tree(t, map[string]string{"project/.git": "", "home/": ""})
	result, err := loadIn(t, root, WithDefaults(map[string]any{"log": map[string]any{"level": "warn"}}))
	if err != nil {
		t.Fatal(err)
	}
	if result.Found {
		t.Fatal("defaults must not set Found")
	}
	if result.Sources[0].Format != "defaults" {
		t.Fatalf("got %#v", result.Sources)
	}
}

func TestLoadThreeLayersAndTheLosersAreStillReported(t *testing.T) {
	root := tree(t, map[string]string{
		"project/.git":                       "",
		"project/.config/mytool/config.toml": "[log]\nlevel = \"debug\"\n",
		"home/.config/mytool/config.toml":    "[log]\nlevel = \"info\"\n",
	})
	result, err := loadIn(t, root, WithEnv(map[string]string{
		"MYTOOL_LOG__LEVEL": "trace",
		"MYTOOL_PORT":       "5432",
	}))
	if err != nil {
		t.Fatal(err)
	}
	want := map[string]any{"log": map[string]any{"level": "trace"}, "port": int64(5432)}
	if !reflect.DeepEqual(result.Config, want) {
		t.Fatalf("got %#v, want %#v", result.Config, want)
	}
	var precedences []int
	for _, source := range result.Sources {
		precedences = append(precedences, source.Precedence)
	}
	if !reflect.DeepEqual(precedences, []int{1, 2, 4}) {
		t.Fatalf("got %v", precedences)
	}
}

func TestLoadOverridesWinOverEverything(t *testing.T) {
	root := tree(t, map[string]string{
		"project/.git":                       "",
		"project/.config/mytool/config.toml": "[log]\nlevel = \"debug\"\n",
		"home/":                              "",
	})
	result, err := loadIn(t, root,
		WithEnv(map[string]string{"MYTOOL_LOG__LEVEL": "trace"}),
		WithOverrides(map[string]any{"log": map[string]any{"level": "silent"}}),
	)
	if err != nil {
		t.Fatal(err)
	}
	if got := result.Config["log"].(map[string]any)["level"]; got != "silent" {
		t.Fatalf("got %v", got)
	}
	if result.Sources[len(result.Sources)-1].Precedence != 5 {
		t.Fatalf("got %#v", result.Sources)
	}
}

func TestLoadRejectsAPackageNameWithASeparator(t *testing.T) {
	if _, err := Load("../evil"); err == nil {
		t.Fatal("expected an error")
	} else if IsKind(err, KindMalformed) {
		t.Fatal("a bad package name is a programming error, not a ConfigError")
	}
}

// SPEC section 5's list is closed, and each entry needs a route that reaches it.
func TestEveryConfigErrorKindIsReachable(t *testing.T) {
	seen := map[Kind]bool{}

	malformedRoot := tree(t, map[string]string{
		"project/.git": "", "project/.config/mytool/config.toml": "[log\n", "home/": "",
	})
	if _, err := loadIn(t, malformedRoot); IsKind(err, KindMalformed) {
		seen[KindMalformed] = true
	} else {
		t.Fatalf("expected malformed, got %v", err)
	}

	duplicateRoot := tree(t, map[string]string{
		"project/.git":                       "",
		"project/.config/mytool/config.yaml": "a: 1\n",
		"project/.config/mytool/config.yml":  "a: 2\n",
		"home/":                              "",
	})
	if _, err := loadIn(t, duplicateRoot); IsKind(err, KindDuplicateFormat) {
		seen[KindDuplicateFormat] = true
	} else {
		t.Fatalf("expected duplicate-format, got %v", err)
	}

	unreadableRoot := tree(t, map[string]string{
		"project/.git": "", "project/.config/mytool/config.toml/": "", "home/": "",
	})
	if _, err := loadIn(t, unreadableRoot); IsKind(err, KindUnreadable) {
		seen[KindUnreadable] = true
	} else {
		t.Fatalf("expected unreadable, got %v", err)
	}

	strictRoot := tree(t, map[string]string{
		"project/.git": "", "project/.config/mytool/config.toml": "extra = 1\n", "home/": "",
	})
	result, err := loadIn(t, strictRoot)
	if err != nil {
		t.Fatal(err)
	}
	var target struct {
		Known string `mapstructure:"known"`
	}
	if err := result.Unmarshal(&target, UnmarshalStrict()); IsKind(err, KindUnknownKey) {
		seen[KindUnknownKey] = true
	} else {
		t.Fatalf("expected unknown-key, got %v", err)
	}

	typeMismatch := tree(t, map[string]string{
		"project/.git": "", "project/.config/mytool/config.toml": "port = \"not a number\"\n", "home/": "",
	})
	result, err = loadIn(t, typeMismatch)
	if err != nil {
		t.Fatal(err)
	}
	var typed struct {
		Port int `mapstructure:"port"`
	}
	if err := result.Unmarshal(&typed); IsKind(err, KindValidation) {
		seen[KindValidation] = true
	} else {
		t.Fatalf("expected validation, got %v", err)
	}

	// not-found exists so a caller can name the condition (SPEC section 5), not so it can be
	// raised: an empty tree returns a nil error.
	emptyRoot := tree(t, map[string]string{"project/.git": "", "home/": ""})
	if _, err := loadIn(t, emptyRoot); err != nil {
		t.Fatalf("nothing found must not raise: %v", err)
	}
	if (&ConfigError{Kind: KindNotFound}).Kind == KindNotFound {
		seen[KindNotFound] = true
	}

	for _, kind := range []Kind{
		KindNotFound, KindUnreadable, KindMalformed,
		KindDuplicateFormat, KindUnknownKey, KindValidation,
	} {
		if !seen[kind] {
			t.Fatalf("kind %q was never reached by a test", kind)
		}
	}
}
